### DISCUSSION TITLE
http2 configuration behind proxy

### DISCUSSION THREAD

**Discussion by timlevett** [grafana/grafana#54592] [Platform]
The idea for this is to decouple the http2 setting in frontend from backend server so that we can leverage frontend http2 sending logic when reverse proxy is configured since the http2 cert would be terminated at the proxy. Right now this is tightly coupled.

What we would really like is 2 things:
1) decouple `http2Enabled` in frontend for backend
2) make [this parameter](https://github.com/grafana/grafana/blob/a74681036e1d455fee0543d1ae1481cee32f655d/public/app/core/services/FetchQueueWorker.ts#L17) configurable so that we can test out more than 1000.

**Comment by sakjur** [comment]
👀 I'm happy to provide any support to this I can from the backend side, this should be a fairly simple thing for us to support.

**Comment by joshhunt** [comment]
It looks like the only place the frontend uses `http2Enabled` is for deciding on the `maxParallelRequests` the fetch queue makes.

If we add in an additional configuration  - something like (naming tbd) `frontend.fetchQueueParallelism` - that'll override the default 1000/5, would that suffice for the frontend?

It could look something like 

```js
const maxParallelRequests = config?.fetchQueueParallelism === 0
  ? (config?.http2Enabled ? 1000 : 5)
  : config?.fetchQueueParallelism;
```

**Reply by zhichli** [reply]
@sakjur @joshhunt thanks for your responses! I guess the proposal by @joshhunt would do the trick for both needs. The only caveat is that it might be necessary to enforce (or at least **inform/warn** in grafana.ini via comment for this configuration) some limits for parallelism ?

**Reply by zhichli** [reply]
Side question about frontend requests sending if @joshhunt you could answer: is it the truth that at any point of time the frontend only sends query requests for panels being rendered in the visible window (so any other queries in invisible panels would not be sent until they get visible after scrolling ? If so, could you please point me to the code that implement this feature ?

**Reply by sakjur** [reply]
@zhichli I think an info message will suffice. Enforced limits should probably be done on the backend using quotas. Since the parallelism here is self-regulation on behalf of the frontend, a malicious actor could trivially bypass it so the problem would mostly be if someone set it too high by mistake, which seem like a tolerable footgun to provide users with.

But something like "`frontend.max_parallel_requests` is configured to 1234 which might cause congestion. Try reducing this value to a value between 5-30 if you're experiencing performance issues when loading dashboards."?

Improved quotas enforced by the backend is something we probably need to develop in a near future. Aaany day now 😅

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
